src/Orleans.Serialization/Buffers/Reader.cs

```diff
@@ -837,6 +837,12 @@ public unsafe uint ReadVarUInt32()
 
                 ulong result = Unsafe.ReadUnaligned<ulong>(ref readHead);
                 var bytesNeeded = BitOperations.TrailingZeroCount((uint)result) + 1;
+
+                if (bytesNeeded > 9)
```

Comment:
Isn't the maximum possible value here 4 (5-1)?
And even then there's some potential for a value that's slightly too large for an uint32 (if the 5 byte value is greater than 0x1FFFFFFFF0), although checking for that would probably be too expensive (masking would need to be changed back to 64bit and adding a checked conversion).

Reply:
Actually, why don't we do the masking before shifting? The mask would be a lot simpler then.

Reply:
> Actually, why don't we do the masking before shifting? The mask would be a lot simpler then.
Could you demonstrate?

Reply:
Feel free to add commits to this PR

Reply:
```cs
var bytesNeeded = BitOperations.TrailingZeroCount((uint)result) + 1;
_bufferPos = pos + bytesNeeded;
result &= (1UL << (bytesNeeded * 8)) - 1;
result >>= bytesNeeded;
return checked((uint)result);